File: src/main/java/com/freya02/botcommands/components/builder/ComponentBuilderImpl.java
package com.freya02.botcommands.components.builder;

import java.util.concurrent.TimeUnit;

@SuppressWarnings("unchecked")
public abstract class ComponentBuilderImpl<T extends ComponentBuilderImpl<T>> implements ComponentBuilder<T> {
	private boolean oneUse;
	private long ownerId;

	private long timeout;

	@Override
	public T oneUse() {
		this.oneUse = true;

		return (T) this;
	}

	@Override
	public T ownerId(long ownerId) {
		this.ownerId = ownerId;

		return (T) this;
	}

	@Override
	public T timeout(long timeout, TimeUnit timeoutUnit) {
		this.timeout = timeoutUnit.toMillis(timeout);

		return (T) this;
	}

	@Override
	public boolean isOneUse() {
		return oneUse;
	}

	@Override
	public long getOwnerId() {
		return ownerId;
	}

	@Override
	public long getTimeout() {
		return timeout;
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/new_components/builder/ComponentBuilderImpl.kt
package com.freya02.botcommands.internal.new_components.builder

import com.freya02.botcommands.api.components.InteractionConstraints
import com.freya02.botcommands.api.new_components.builder.ComponentBuilder

@Suppress("UNCHECKED_CAST")
abstract class ComponentBuilderImpl<T: ComponentBuilder<T>> : ComponentBuilder<T> {
    final override var oneUse: Boolean = false
        private set
    final override var constraints: InteractionConstraints = InteractionConstraints()
        private set

    override fun oneUse(): T = this.also { oneUse = true } as T

    override fun constraints(block: InteractionConstraints.() -> Unit): T = this.also { constraints.apply(block) } as T

    override fun setConstraints(constraints: InteractionConstraints): T = this.also { this.constraints = constraints } as T
}